### ISSUE TITLE
[network] Terminate inbound Noise handshake early when pubkey not in elligible peers set

### ISSUE THREAD

**Issue by phlip9** [libra/libra#1400]
# 🚀 Feature Request

Validators currently use authenticated and encrypted Noise channels to communicate with each other. In addition, when running as a Validator, we reject all connections that authenticate to a pubkey outside of the validator set (https://github.com/libra/libra/blob/master/network/src/transport.rs#L94). However, we only reject the connection after (1) fully negotiating the Noise protocol and (2) completing the full Noise handshake. For context, our current logic looks like:

```
# Protocol Negotiation
> /libra/protocol-interactive/1.0.0
> /noise_ix_25519_aesgcm_sha256/1.0.0
< /noise_ix_25519_aesgcm_sha256/1.0.0

# Noise_IX handshake
> e, s
< e, ee, se, s, es

# Reject unathenticated peers
if s not in validator_set { disconnect }

# ..
```

However, as an _inbound_ listener, we can reject a connection as soon as we read the static pubkey, meaning we can save ourselves resources by not responding to the Noise handshake. This helps mitigate a potential DDoS vector where non-validators flood our inbound with slow connections that, while failing to authenticate, still waste our CPU + bandwidth. Our inbound protocol would then look like:

```
# Protocol Negotiation
> /libra/protocol-interactive/1.0.0
> /noise_ix_25519_aesgcm_sha256/1.0.0
< /noise_ix_25519_aesgcm_sha256/1.0.0

# Noise_IX handshake
> e, s                                  # if s not in validator_set { disconnect }
< e, ee, se, s, es

# ..
```

One issue with Noise IX is that a peer can still force our response by sending a real validator's `s` static pubkey in the first message. To combat this attack, we should move to Noise IK, which allows us to verify that the dialer actually controls the private key for `s` in the first message and then disconnect accordingly:

```
# Noise_IK handshake
> e, es, s, ss                          # if s not in validator_set { disconnect }
< e, ee, se
```

As a further optimization, we can remove the initial roundtrip from protocol negotiation with #1158. Our inbound protocol would approach optimal:

```
# Protocol Negotiation
> /libra/protocol-select/1.0.0
> /noise_ik_25519_aesgcm_sha256/1.0.0
# Noise_IK handshake
> e, es, s, ss                          # if s not in validator_set { disconnect }
< e, ee, se

# ..
```

cc @bothra90 @bmwill

**Comment by mimoo** [comment]
https://github.com/libra/libra/pull/3951 terminates the connection after receiving the first handshake message if it does not recognize the key (if in the validator network).

For context, the PR moves from IX to IK:

    IK:
      <- s
      ...
      -> e, es, s, ss
      <- e, ee, se

We could in theory abort the connection right after `s` and before the additional Diffie-Hellman key exchange `ss`, but note that this would mean modifying the behavior of the noise implementation. If others want to implement this, and rely on a noise library, they will have to modify the library's behavior as well.

In addition, if someone really wants to fuck with us, they can still send us a correct validator static key `s` at this point. We can only realize this after doing `ss` and noticing that our derived secret doesn't match theirs (because they can't perform the key exchange).

For these reasons I don't think it makes sense to abort the handshake right after receiving the `s` key.

**Comment by phlip9** [comment]
Closed by #3951

As described in the PR, we only decide to drop the inbound upgrade after fully parsing the client message. While this is slightly more work for the server than is optimal, it's also simpler and easier to maintain. Either way, we can avoid computing the response message for an unauthenticated peer and drop the connection immediately without sending any response bytes over-the-wire.
